### PR TITLE
Unit test doc

### PR DESCRIPTION
**Pull request by Sriep**
Document to help standardise 0chian unit tests.  Please suggest input.

### CODE REVIEW

File: docs/unit-tests.md [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+### Vektra mockery
+
+To help with our unit tests, in `0chain`, we autogenerate mock object for all our
+interface objects using the
+[vektra/mockery](https://github.com/vektra/mockery) package. 
+
+As we autogenerate these mock object, we avoid keeping them under
+version control. Instead, we generate them when we run unit tests during 
+continuous integration. Developers can generate them using the 
+[0chain\generate_mcoks.sh](https://github.com/0chain/0chain/blob/master/generate_mocks.sh) script.
```
Comment (shravanshetty1) [inline docs/unit-tests.md:289]: typo

**Review by shravanshetty1 — CHANGES_REQUESTED** [review]

**Comment by Dmdv** [comment]
@Sriep Do we use Ginkgo and gomock or it is not encouraged?

**Comment by Dmdv** [comment]
@Sriep   
I found no mentions about a convention where to keep all test factory code.  
Imagine not interfaces but types which are not generated by **mockery**.  
Some use-test cases are repetitive throughout the same category tests and factory for said types are definitely worth keeping in the separate directory rather than being kept in a single file.

Keeping all test types factories make the code clumpy

**Comment by Sriep** [comment]
```
Do we use Ginkgo and gomock or it is not encouraged?
```
Don't use Ginko or gomock unless we decide to use gomock or Ginkgo.

**Comment by Sriep** [comment]
```
I found no mentions about a convention where to keep all test factory code.
```
How do you mean `test factory code?` We want to keep tests decoupled where possible. 

Some common functions maybe, but we want to avoid a large amount of common setup code 
that is not necessary for any particular test.

**Comment by Dmdv** [comment]
> ```
> I found no mentions about a convention where to keep all test factory code.
> ```
> 
> How do you mean `test factory code?` Keep tests decoupled where possible. Also, try to avoid a large amount of common setup code that is not necessary for any particular test.

@Sriep 
Do you mean keeping all tests outside of the package folder in the test folder or keeping all tests alongside the tested files?

As for test factory code, I mean some common types created in different tests with copy-pasting. 
Do we place them in a separate folder?

**Comment by Sriep** [comment]
For common methods and objects just be sensible. A common help _test file is fine.

**Pull request merged**